src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -25,16 +24,21 @@ private sealed class StateSet
             public readonly string Language;
             public readonly DiagnosticAnalyzer Analyzer;
 
-            private readonly ConcurrentDictionary<DocumentId, ActiveFileState> _activeFileStates;
-            private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectStates;
+            private ImmutableDictionary<DocumentId, ActiveFileState> _activeFileStates;
+            private ImmutableDictionary<ProjectId, ProjectState> _projectStates;
+
+            // Used to protect writes of _activeFileStates or _projectStates
+            private readonly object _lock;
```

Comment:
if we have a lock, do we even need the dictionaries to be immutable?  could we just use normal dictionaries?

Reply:
ah, i see below you often will just grab the variable, and party on it without a lock.  immutable dict seems fine then.